### ISSUE TITLE
Formatting error

### ISSUE THREAD

**Issue by JackKenney** [BetterThanTomorrow/calva-fmt#4]
If I type:
```
(defn foo
  [x]
  (+ x
     (* x
        1)
     x
     x
x
```
I would expect the formatter to reformat like this:
```
(defn foo
  [x]
  (+ x
     (* x
        1)
     x
     x
     x
```
but instead I get the following error:
```
Running the contributed command:'calva-fmt.formatCurrentForm' failed
```
Is this a bug I can submit for review?

**Comment by PEZ** [comment]
It could be a feature request at least. But a tricky one to address.

_TLDR; You'll need to help Calva Formatter with where the forms end._

The reason it fails is that the form is not complete. Judging from how you would expect it to be formatted I think it should look like so:

```clojure
(defn foo
  [x]
  (+ x
     (* x
        1)
     x
     x
     x))
```

But Calva Formatter does not infer the missing parens and then the formatting engine croaks on what is not valid lisp syntax. This inference is tricky to do, since you could add the missing parens on many different places. In fact the only hint of where they should be is the indentation. I think that [Parinfer](https://shaunlebron.github.io/parinfer/) would place them like so:

```clojure
(defn foo
  [x]
  (+ x
     (* x
        1)
     x
     x))
x
```

And then Calva Formatter would think it was already nicely formatted.

**Comment by JackKenney** [comment]
Gotcha. It's a tricky one for sure, but it's the expectation that a lot of Lisp programmers have and it keeps them using things like EMACS instead of more robust GUIs like vscode.

In certain editors the standard is to indent everything as far in as it can go unless closing parentheses are present and un-indent them if they are.

I can look into submitting it as a feature request.

Thanks!

**Comment by lspector** [comment]
I was about to post an issue "Formatter fails on unbalanced forms" when I saw this discussion. 

I think that the standard approach is to assume everything will be closed at the end.

This produces the correct behavior while one is writing an expression, if there are not already a complete set of closing brackets. This can happen as a result of nonlinear editing or cut/paste, etc. In my own coding this is a common occurrence.

When using the formatter to help spot bugs (very common in my teaching experience), the brackets are more likely to be unbalanced, and assuming they'll all be closed at the end is the way to see the problems.

So it'd be great to have this feature implemented, with the assumtion that everything will be closed at the end.

FWIW my opinion is that doing any of this based on indentation, as with parinfer, would be seriously counterproductive. The point of the formatter is to make the format reflect the structure, not the other way around.

**Comment by PEZ** [comment]
Calva has gotten better at always inserting closing brackets together with the opening ones, as well as keeping the code correctly indented as it is typed in, so this situation should happen less often. Reopening this anyway so that I can better understand the use cases and gauge how hard it would be to solve.

**Comment by lspector** [comment]
FWIW I'm actually not crazy about the insertion of closing brackets together with the opens, since I often want an open without a close (because I'm going to insert the close somewhere else), and this behavior means I have to explicitly delete the close every time I type an open. I often observe students having to do this as well.

This isn't so terrible, however, and I know that a lot of people like getting the closes automatically. I've dealt with several editors that do this, and I think people can work around it.

However, whether closes are added automatically or not, if you allow freeform editing and cut/paste (which I do think are both essential) then sometimes the brackets will become mismatched. One of the best ways to see what's going on in such cases is to invoke the automatic formatting. However, this won't work if the automatic formatting only works if brackets are already balanced.

Fortunately, I think the solution is straightforward: The formatter should act as if there are enough closes at the end of the expression to balance any excess of opens, and ignore any extra closes. This makes everything format correctly for partially written expressions, and allows one to see the structure of code that is incomplete or imbalanced.

This is what is done in several other Lisp editors, including at least some Emacs setups, Gorilla REPL, Counterclockwise, and probably others.

**Comment by PEZ** [comment]
In theory you should be able to override the auto-closing behaviour, but that doesn't work when I try it. However, for me, having them always entered in pairs helps me a lot. I often enter a paren before a form I want to enclose and then slurp the form using Paredit.

But back to the original issue. Thanks for the help with thinking about it. Closing everything at the end of the expressions seems straightforward, but asks for a few more things to be solved:

* Where is the end of the expression? End of the file? End of something else? What do the other editors do here? Calva Formatter currently has ways to find things like the top level form, but it always relies on that the forms are complete.
* Related. Calva Formatter avoids the non-trivial task of counting brackets by relying on structural ways to navigate the code. This breaks down if the code is not structurally intact. I have had code doing this counting in several variants and could do it that way again for this use case, but it will mean I have to solve some issues with my previous attempts (it is mainly about skipping brackets inside strings and ”literal” comments).

**Comment by lspector** [comment]
> Where is the end of the expression? End of the file? End of something else? What do the other editors do here?

In the others with which I am familiar the formatter is mostly invoked on the current selection (or just the line containing the insertion point if there is no selection), and the end of the expression is the end of the selection/line. When you want to clean up or debug an expression, you select it and hit the key(s) to format, and it just reformats the stuff that was selected, assuming everything is closed at the end. If you select the whole file, or in some cases invoke a version of the command that automatically applies to the whole file rather than a selection, then it will reformat everything on the assumption that everything is closed up at the end.

If Calva is always being invoked on the whole file then I think assuming everything is closed at the end is the way to go. If an expression is missing a close, then everything below it will be indented too far, but that's fine because it will provide the right visual clue to where the problem is. Once the missing brackets are added, everything below it will move back to where it should be.

> Related. Calva Formatter avoids the non-trivial task of counting brackets...

I am not sure how it is implemented in other systems, but roughly the same approach has been taken in Lisp editors I've used going back to the 1980s, and I suspect that there are some shared bits of code running through several of them. If I recall correctly, Gorilla REPL's author actually didn't know that his system did re-formatting -- and it does it beautifully! -- until we had a conversation about it and he realized that it could be invoked with shift-tab. It's implemented in some library that he uses, and I think that it does exactly the right thing.

> In theory you should be able to override the auto-closing behaviour, but that doesn't work when I try it. However, for me, having them always entered in pairs helps me a lot. I often enter a paren before a form I want to enclose and then slurp the form using Paredit.

Again, this is less of an issue for me than formatting incomplete expressions, but to be more clear about my perspective here, I find any approach involving Paredit to be more of a step backwards than forwards. My view/hope is for people to be able to use their existing typing/editing/cutting/pasting skills, without having to learn new concepts or key commands, or having to undo things that the editor is doing in an attempt to be helpful. I think that re-formatting on demand is essential, both for cleaning things up and spotting structural mistakes. But aside from that, I think the ideal thing is for the editor to act just like any other text editor, that the user already knows how to use and won't be surprised or frustrated by.

In part my perspective comes from the fact that I spend a lot of time teaching students who are either new to programming entirely, or at least new to Lisp-like languages. I really don't want them to have to learn new editing concepts and skills before they've learned the programming concepts and skills. But it's not just about my newbie students, since for my own coding as well I have the same preferences, for an editor that behaves just like every other text editing environment that I use, but for which there's a re-format command (and bracket matching, and maybe some other sweet bells and whistles like popup documentation, etc., which Calva has and which are helpful without getting in the way of editing).

**Comment by mseddon** [comment]
This issue *ought* to be fixable with the new auto-indent logic. This is mostly because now we have a way to search 'up' out of the enclosing sexpr at the lexical, rather than sexpr level. This means we can wind upwards and backwards until we find the toplevel form, and use the parens collected to insert anything that was missing in order to appease `cljfmt`.

There is a subtle issue in that our matcher is slightly less strict and allows mismatched parenthesis types, so the edge case here is to ensure that we correct any mismatched closing parens that would cause cljfmt to bail on reading, as well as just auto-closing any remaining parens on the stack.

**Comment by PEZ** [comment]
Archiving this repo now, please open any remaining issues against the [Calva repository](/BetterThanTomorrow/calva).
